internal/core/operations/switch_active_version/switch_active_version_executor.go

```diff
@@ -0,0 +1,197 @@
+//    replace goroutines channel;
+// 4. Switch the active version
+func (e *SwitchActiveVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Debug("start switching scheduler active version")
+
+	updateDefinition := definition.(*SwitchActiveVersionDefinition)
+	scheduler := &updateDefinition.NewActiveScheduler
+
+	maxSurgeNum, err := e.roomManager.SchedulerMaxSurge(ctx, scheduler)
+	if err != nil {
+		return fmt.Errorf("error fetching scheduler max surge: %w", err)
+	}
+
+	roomsChan := make(chan *game_room.GameRoom)
+	var maxSurgeWg sync.WaitGroup
+
+	maxSurgeWg.Add(maxSurgeNum)
+	for i := 0; i < maxSurgeNum; i++ {
+		go e.replaceRoom(logger, &maxSurgeWg, roomsChan, e.roomManager, *scheduler)
+	}
+
+roomsListLoop:
+	for {
+		rooms, err := e.roomManager.ListRoomsWithDeletionPriority(ctx, scheduler.Name, scheduler.Spec.Version, maxSurgeNum, e.roomsBeingReplaced)
+		if err != nil {
+			return fmt.Errorf("failed to list rooms for deletion")
+		}
+
+		for _, room := range rooms {
+			e.roomsBeingReplaced.Store(room.ID, true)
+			select {
+			case roomsChan <- room:
+			case <-ctx.Done():
+				break roomsListLoop
+			}
+		}
+
+		if len(rooms) == 0 {
+			break
+		}
+	}
+
+	// close the rooms change and ensure all replace goroutines are gone
+	close(roomsChan)
+	maxSurgeWg.Wait()
+
+	err = e.schedulerManager.SwitchActiveScheduler(ctx, scheduler)
+	if err != nil {
+		logger.Error("Error switching active scheduler version on scheduler manager")
+		return err
+	}
+
+	logger.Debug("scheduler update finishes with success")
+	return nil
+}
+
+func (e *SwitchActiveVersionExecutor) OnError(ctx context.Context, op *operation.Operation, definition operations.Definition, executeErr error) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Info("starting OnError routine")
+
+	err := e.deleteNewCreatedRooms(ctx, logger)
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("finished OnError routine")
+	return nil
+}
+
+func (e *SwitchActiveVersionExecutor) Name() string {
+	return OperationName
+}
+
+func (e *SwitchActiveVersionExecutor) replaceRoom(logger *zap.Logger, wg *sync.WaitGroup, roomsChan chan *game_room.GameRoom, roomManager *room_manager.RoomManager, scheduler entities.Scheduler) {
+	defer wg.Done()
+
+	// we're going to use a separated context for each replaceRoom since we
+	// don't want to cancel the replace in the middle (like creating a room and
+	// then left the old one (without deleting it).
+	ctx := context.Background()
+
+	for {
+		room, ok := <-roomsChan
+		if !ok {
+			return
+		}
+
+		gameRoom, _, err := roomManager.CreateRoom(ctx, scheduler)
+		if err != nil {
```

Comment:
@caiordjesus The current code won't make the operation fail if we have an error creating the new game room, I think we need to change this logic to make the execution return with an error when this happens, right?